cadc-inventory/src/main/java/org/opencadc/inventory/Artifact.java

```diff
@@ -0,0 +1,189 @@
+    
+    public String contentType;
+    public String contentEncoding;
+    
+    /**
+     * Constructor.
+     * 
+     * @param uri logical identifier
+     * @param contentChecksum checksum of the content, form: {algorithm}:{hexadecimal value}
+     * @param contentLastModified last-modified timestamp of the content
+     * @param contentLength number of bytes in the content
+     */
+    public Artifact(URI uri, URI contentChecksum, Date contentLastModified, Long contentLength) {
+        InventoryUtil.assertNotNull(Artifact.class, "uri", uri);
+        InventoryUtil.assertNotNull(Artifact.class, "contentChecksum", contentChecksum);
+        InventoryUtil.assertNotNull(Artifact.class, "contentLastModified", contentLastModified);
+        InventoryUtil.assertNotNull(Artifact.class, "contentLength", contentLength);
+        if (contentLength <= 0L) {
+            throw new IllegalArgumentException("invalid " + Artifact.class.getSimpleName() + ".contentLength: " + contentLength);
+        }
+        this.uri = uri;
+        this.contentChecksum = contentChecksum;
+        this.contentLastModified = contentLastModified;
+        this.contentLength = contentLength;
+        this.uriBucket = computeBucket(uri);
+    }
+
+    /**
+     * @return the logical identifier
+     */
+    public URI getURI() {
+        return uri;
+    }
+    
+    /**
+     * @return five character bucket tag for batch operations
+     */
+    public String getBucket() {
+        return uriBucket;
+    }
+    
+    /**
+     * Content checksum is a URI of the form {algorithm}:{hexadecimal value}.
+     * 
+     * @return the content checksum
+     */
+    public URI getContentChecksum() {
+        return contentChecksum;
+    }
+
+    /**
+     * @return the last modifidied timestamp of the content
+     */
+    public Date getContentLastModified() {
+        return contentLastModified;
+    }
+
+    /**
+     * @return size of the content in bytes
+     */
+    public Long getContentLength() {
+        return contentLength;
+    }
+    
+    private String computeBucket(URI uri) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-1");
+            byte[] bytes = super.primitiveValueToBytes(uri, "File.uri", md.getAlgorithm());
+            md.update(bytes);
+            byte[] sha = md.digest();
+            String hex = HexUtil.toHex(sha);
+            return hex.substring(0, 5); // 16^5 = 1 million buckets
+        } catch (NoSuchAlgorithmException ex) {
+            throw new RuntimeException("BUG: failed to get instance of SHA-1", ex);
+        }
+    }
+
+    /**
+     * Compares uri values.
+     * 
+     * @param o object to compare to
+     * @return true if the logical identifiers are equal, otherwise false
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (o == null) {
+            return false;
+        }
+        Artifact f = (Artifact) o;
+        return uri.equals(f.uri);
```

Comment:
I remember there being something strange about URI equals() but don't recall what...

Reply:
UUID.compareTo in java doesn't comply with the UUID spec. Maybe that was it?

Reply:
Hmm, maybe.  Javadoc for URI.equals() seems like it's doing the right thing anyhow.